### ISSUE TITLE
ci failing due to rosidl_generator_py extra parenthesis

### ISSUE THREAD

**Issue by flynneva** [ros-realtime/reference-system#11]
creating this issue to track the problem and offer the solution that worked for me until [this fix gets merged in](https://github.com/ros2/rosidl_python/pull/144) and released.

The CI / local builds will fail due to the way we wrote [our message idl file](https://github.com/ApexAI/reference-system-autoware/blob/main/reference_interfaces/msg/TransmissionStats.idl#L8) by using `char node_name[56]`. Essentially there was a bug in the `rosidl_generator_py` package that causes an extra parenthesis to be added to a build file that is then used by anything that uses that interface message type.

Two solutions present themselves: change the type in the message idl or wait for the fix to be applied.

If we would rather wait for the change, the way to get this all working is to [apply the fix](https://github.com/ros2/rosidl_python/pull/144) to the generated file in your local `colcon_ws/build` directory.  All you need to do is change one of the last lines in the following file:

``` diff
# `build/reference_interfaces/rosidl_generator_py/reference_interfaces/msg/_transmission_stats.pyL167
- all(val >= 0 and val) < 256 for val in value)), \
+ all(ord(val) >= 0 and ord(val) < 256 for val in value)), \

**Comment by flynneva** [comment]
i went ahead and tried various different types like `string node_name[56]` and even `string<56> node_name` and they both compile, with warnings, but cannot run after building.

looks to me like the way we did it with `char node_name[56]` is going to be the most straight forward approach.

**Comment by flynneva** [comment]
as we dont really know when that rosidl_generator_py fix will get implemented, I wanted to make it a bit easier for others to actually build/use these packages. this most recent commit should allow for users to simply "rebuild" the reference_interfaces package and cmake should do the rest and fix the build file for them.

still figuring out how to automatically do this without having to rebuild so we could add it to the CI...but maybe save that for another day or if someone else has some ideas.

when you build for the first time locally it should warn you of the bug and offer the solution:

colcon build
Starting >>> reference_interfaces
--- stderr: reference_interfaces                              
WARN Generated IDL file has an error, rerun `colcon build` to fix the generated file automatically
---
Finished <<< reference_interfaces [3.46s]
Starting >>> reference_system
Finished <<< reference_system [1.33s]                
Starting >>> autoware_reference_system
Finished <<< autoware_reference_system [8.99s]                      

Summary: 3 packages finished [13.9s]
  1 package had stderr output: reference_interfaces
next time you build the CMake file will alter the generated build file with the fix automatically.

**Comment by flynneva** [comment]
was able to solve this issue rather cleanly by adding this to the `reference_interfaces` package `CMakelists.txt` file which edits the generated file if it finds the line that causes the error as mentioned above:

```
# fix rosidl_generator_py bug #143
# https://github.com/ros2/rosidl_python/issues/143
set(GENERATED_FILE "${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_py/${PROJECT_NAME}/msg/_transmission_stats.py")

message(STATUS "checking generated file: ${GENERATED_FILE}")
add_custom_command(
  TARGET ${PROJECT_NAME}__python
  POST_BUILD
  COMMAND sed -i "s/all(val >= 0 and val) < 256/all(ord(val) >= 0 and ord(val) < 256/" ${GENERATED_FILE}
  COMMENT "Check generated IDL files for extra parenthesis..."
  VERBATIM)
  ```
